api/src/controllers/changes.js

```diff
@@ -1,578 +1,396 @@
-var _ = require('underscore'),
-    async = require('async'),
-    auth = require('../auth'),
-    config = require('../config'),
-    serverUtils = require('../server-utils'),
-    db = require('../db-nano'),
-    ALL_KEY = '_all', // key in the docs_by_replication_key view for records everyone can access
-    UNASSIGNED_KEY = '_unassigned', // key in the docs_by_replication_key view for unassigned records
-    CONTACT_TYPES = ['person', 'clinic', 'health_center', 'district_hospital'],
-    inited = false,
-    continuousFeeds = [];
-const uuid = require('uuid/v4');
-const heartbeatFilter = require('../services/heartbeat-filter');
-
-/**
- * As per https://issues.apache.org/jira/browse/COUCHDB-1288, there is a 100 doc
- * limit on processing changes feeds with the _doc_ids filter within a
- * reasonable amount of time.
- */
-const MAX_DOC_IDS = 100;
-
-var error = function(code, message) {
-  return JSON.stringify({ code: code, message: message });
-};
-
-var getDepth = function(userCtx) {
-  if (!userCtx.roles || !userCtx.roles.length) {
-    return -1;
+const auth = require('../auth'),
+      db = require('../db-pouch'),
+      authorization = require('../services/authorization'),
+      _ = require('underscore'),
+      config = require('../config'),
+      serverUtils = require('../server-utils'),
+      heartbeatFilter = require('../services/heartbeat-filter'),
+      tombstoneUtils = require('@shared-libs/tombstone-utils')(Promise, db.medic),
+      uuid = require('uuid/v4');
+
+let inited = false,
+    continuousFeed = false,
+    longpollFeeds = [],
+    normalFeeds = [],
+    MAX_DOC_IDS = 100,
+    currentSeq = 0;
+
+const DEFAULT_TIMEOUT = 60000,
+      DEFAULT_LIMIT = 100,
+      DEBOUNCE_INTERVAL = 200;
+
+const split = (array, count) => {
+  if (count === null || count < 1) {
+    return array;
   }
-  var settings = config.get('replication_depth');
-  if (!settings) {
-    return -1;
+  const result = [];
+  let i = 0,
+      length = array.length;
+  while (i < length) {
+    result.push(array.slice(i, i += count));
```

Comment:
I think it would be easier to read if you used `splice` instead of `slice`, eg:

```js
  const result = [];
  while (array.length) {
    result.push(array.splice(0, count));
  }
```

Reply:
Sure, easier to read, but also mutating. 
I do use splice now, and save the "split" result.